LongestPole/Main.lean

```diff
@@ -0,0 +1,191 @@
+
+structure SpeedCenterSource_1 where
+  repo_id : String
+  hash : String
+deriving ToJson, FromJson
+
+structure SpeedCenterSource_2 where
+  source : SpeedCenterSource_1
+deriving ToJson, FromJson
+
+structure SpeedCenterDimension where
+  benchmark : String
+  metric : String
+  unit : String
+deriving ToJson, FromJson
+
+structure SpeedCenterMeasurement where
+  dimension : SpeedCenterDimension
+  value : Float
+deriving ToJson, FromJson
+
+structure SpeedCenterResult where
+  measurements : List SpeedCenterMeasurement
+deriving ToJson, FromJson
+
+structure SpeedCenterRun_3 where
+  id : String
+  source : SpeedCenterSource_2
+  result : SpeedCenterResult
+deriving ToJson, FromJson
+
+structure SpeedCenterRunResponse where
+  run : SpeedCenterRun_3
+deriving ToJson, FromJson
+
+structure SpeedCenterMessage where
+  repo_id : String
+  message : String
+  commit_hash : String
+deriving ToJson, FromJson
+
+def SpeedCenterRunResponse.instructions (response : SpeedCenterRunResponse) :
+    NameMap Float := Id.run do
+  let mut r : NameMap Float := ∅
+  for m in response.run.result.measurements do
+    let n := m.dimension.benchmark
+    if n.startsWith "~" then
+      r := r.insert (n.drop 1).toName m.value
+  return r
+
+def speedCenterRunResponse (run : String) : IO SpeedCenterRunResponse := do
+  let r ← speedCenterRunJson run
+  match Json.parse r with
+  | .error e => throw <| IO.userError s!"Could not parse speed center JSON: {e}\n{r}"
+  | .ok j => match fromJson? j with
+    | .ok v => pure v
+    | .error e => match fromJson? j with
+      | .ok (v : SpeedCenterMessage) =>
+        IO.eprintln s!"http://speed.lean-fro.org says: {v.message}"
+        IO.eprintln s!"Try moving to an older commit?"
+        IO.Process.exit 1
+      | .error _ => throw <| IO.userError s!"Could not parse speed center JSON: {e}\n{j}"
+
+def headSha : IO String := return (← runCmd "git" #["rev-parse", "HEAD"]).trim
+
+def speedCenterInstructions (run : String) : IO (NameMap Float) :=
+  return (← speedCenterRunResponse run).instructions
+
+partial def cumulativeInstructions (instructions : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Float :=
+  graph.fold (init := ∅) fun m n _ => go n m
+where
+  go (n : Name) (m : NameMap Float) : NameMap Float :=
+    if m.contains n then
+      m
+    else
+      let parents := (graph.find? n).getD #[]
+      let m := parents.foldr (init := m) fun n' m => go n' m
+      let t := (parents.map fun n' => (m.find? n').getD 0).foldr max 0
+      m.insert n ((instructions.find? n).getD 0 + t)
+
+def slowestParents (cumulative : NameMap Float) (graph : NameMap (Array Name)) :
+    NameMap Name :=
+  graph.fold (init := ∅) fun m n parents =>
+    match parents.toList with
+    | [] => m
+    | h :: t => Id.run do
+      let mut r := h
+      for n' in t do
+        if (cumulative.find? n').getD 0 > (cumulative.find? r).getD 0 then
```

Comment:
Is there a reason to prefer `.getD 0` over `.get!`? These lines become quite long... I could imagine even having a `.find!`.

(Also elsewhere.)

Reply:
Seems like some of these `find?` can be replaced by `find!` and some can't (`instructions` doesn't have all the names that the other `nameMap`s have). I'll replace the ones I can.

Reply:
I spotted a PANIC in a recent run of the code, so I must have made a mistake with this somewhere

Reply:
side remark: I believe `(cumulative.find? n).getD 0` can be golfed to `cumulative.findD n 0`